feat(audio): mute on pause

mwerle opened the pull request:
Mute all audio when the game is paused, either by opening the settings window, or by clicking on the Pause icon, and resume when the game is resumed.

NOTE: This PR conflicts directly with #5644 ; so as it appears this is very much personal preference, it may be necessary to add a settings toggle which would make this a slightly less trivial PR.

<!-- Please describe new feature, possible with screenshot if needed. -->
<!-- Please make sure you've read documentation on contributing -->

DamagedEngine commented:
The toggle could have a third option added which only mutes the game when the game window is out-of-focus.

The setting could be called "Mute on pause:" with the following options:

1. Always
2. When out-of-focus
3. Never


This way the music would not disturb players who pause the game to do something else, without interrupting the music for players who only paused the game to interact with menus and look around.

impaktor commented:
I would expect music to keep playing in paused state.

mwerle commented:
> I would expect music to keep playing in paused state.

I would not - hence this PR.  Especially things like engine noise are incredibly annoying (to me) when I pause the game (I almost always have the music muted already anyway) to work on tweaking some code, but the audio engine does not allow just pausing and resuming the FX (intended or bug?)

However, point taken, I will create an option. Not sure if it's (easily) possible to detect the out-of-focus state, but will check if that can be supported.

Thank you for the feedback.

mwerle commented:
> Perhaps it would be better to just add a mute-on-defocus option without tying it to the pause function? I do think the game needs a way to stop it from playing sound in the background when another window such as a web browser is in focus. That desire is presumably why some users might expect the game to become muted on pause.

Conversely I personally do NOT want the audio to become muted just because I switch to a note-taking app or browser in the middle of a gaming session when I am not pausing the game to do so (eg, during a long transfer burn). I am still playing, just using out-of-game tooling to support my game experience.

As I said, this is highly personal. On the one hand having a million game option toggles should be avoided (difficult to test, if nothing else), otoh, something as personal as this I think warrants it.

mwerle commented:
> IMO, the focus-related events should be pushed to the UI event queue rather than the main game event queue. ... There is an existing example of how to do so in the codebase; you can find it by looking up callsites of the appropriate LuaEvent overloads which take a LuaRef as their first argument.

Based on the only example I could find ("onGamePaused" and "onGameResumed"), I changed the event emitting to:
```C
				LuaEvent::Queue(PiGui::GetEventQueue(), hasFocus? "onFocusGained" : "onFocusLost");
```

but haven't figured out how to register and receive these events on the LUA side, as "onGamePaused" is _also_ emitted to the main game event queue and registered in the normal way on the LUA side.

sturnclaw commented on data/pigui/views/game.lua:
```diff
@@ -277,6 +279,37 @@ Event.Register("onPauseMenuClosed", function()
 	Input.EnableBindings()
 end)
 
+Event.Register("onGamePaused", function()
```
These four event handlers should become `ui.Events:Register(..., function() ... end)`. `onGamePaused` et. al. is pushed to the UI event queue and there is no guarantee that events on the main queue will be processed until the pause ends.

sturnclaw commented on src/Pi.cpp:
```diff
@@ -1107,6 +1107,21 @@ void GameLoop::Update(float deltaTime)
+			bool hasFocus = (flags & SDL_WINDOW_INPUT_FOCUS) != 0;
+			if (hasFocus != lastFocus) {
+				lastFocus = hasFocus;
+				LuaEvent::Queue(hasFocus? "onFocusGained" : "onFocusLost");
```
The proper place to handle this would be to listen for `SDL_WINDOWEVENT_FOCUS_GAINED` et. al. in `core/GuiApplication.cpp`, specifically the PollEvents function. That should dispatch to a virtual member method:

```cpp
virtual void OnWindowFocusChanged(bool newFocus) {}
```

That method would then be implemented by `Pi::App`, and an event dispatched to the UI event queue like so:

```cpp
void Pi::App::OnWindowFocusChanged(bool newFocus)
{
    if (!PiGui::GetEventQueue().IsValid())
        return;
    LuaEvent::Queue(PiGui::GetEventQueue(), newFocus ? "onFocusGained" : "onFocusLost");
}
```

mwerle replied:
Thank you; implemented in b5e8818e7474e208ba8bd01a13dc823317f18737 (to be squashed)

sturnclaw commented on src/lua/LuaEngine.cpp:
```diff
@@ -775,6 +775,39 @@ static int l_engine_set_music_volume(lua_State *l)
+	if (lua_isnone(l, 1)) {
+		return luaL_error(l, "SetMuteOnPause takes one boolean argument");
+	}
+	const bool mute= lua_toboolean(l, 1);
```
Nitpick: the codebase's stylistic conventions require a space on both sides of the equals-sign here. We used to have a clang-format nag running on PRs to catch this stuff automatically, but it got far too opinionated without config-file levers to disable its opinions.

sturnclaw commented on data/pigui/modules/settings-window.lua:
```diff
@@ -340,6 +340,8 @@ local function showSoundOptions()
 	local effectsMuted = Engine.GetEffectsMuted()
 	local effectsLevel = Engine.GetEffectsVolume()*100
+	local muteOnPause = Engine.GetMuteOnPause()
+	local muteOnFocus = Engine.GetMuteOnFocus()
```
On a tangent: I hate boilerplate like this and would much prefer we just expose an `Engine.GetConfigOption(key, defaultValue)`, `Engine.SetConfigOption(key, value)` API with consistent (and searchable!) string keys used directly.

This is not actionable feedback within the scope of this PR, but is something you could potentially implement in a separate, later PR if you so desired.

mwerle replied:
Not a LUA programmer so will take your lead on this. Personally I'm not so fond of string parameterisation like that since compilers/analysers cannot detect errors, which can lead to some very subtle and difficult to trace bugs.

sturnclaw commented on data/lang/ui-core/en.json:
```diff
@@ -1735,6 +1735,14 @@
   },
+  "MUTE_ON_FOCUS": {
+    "description": "Setting for muting audio when the game window loses focus.",
+    "message": "Mute on Focus"
```
```suggestion
    "message": "Mute on Window Focus Lost"
```

mwerle replied:
Fixed in  3bd058eca69e9ab8b578c0fcff481dd21a67f50e (to be squashed)

mwerle commented:
Thank you for the detailed information.

Addressed all issues, but still need to implement muting only the SFX. That will require a bugfix to the way the SFX are muted/unmuted (SFX volume is not restored when SFX are unmuted).
